### PR TITLE
drivers: i2c: stm32: Simplify PM device handling

### PR DESCRIPTION
**Pull request by erwango**
Some changes to simplify PM device handling in STM32 I2C driver.

### CODE REVIEW

**Comment by erwango** [comment]
@knthm, when you have time

File: drivers/i2c/i2c_ll_stm32.c
Comment (knthm) [inline drivers/i2c/i2c_ll_stm32.c:76]: I generally agree with your sentiment, pinctrl isn't necessary to configure the bus.
However, the `pm_device_runtime_get/put` functions also operate as a sort of semaphore with the usage field, so if the usage counter is already >1 when we reach this section, we'll unnecessarily be toggling the clock where the runtime get/put wouldn't.

File: drivers/i2c/i2c_ll_stm32.c
Comment (knthm) [inline drivers/i2c/i2c_ll_stm32.c:395]: If the `i2c_stm32_runtime_configure` would retain the `pm_device_runtime_get/put` logic it has currently, it would make more sense to remove only `i2c_stm32_suspend` and leave `pm_device_init_suspended`

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:76]: > so if the usage counter is already >1 when we reach this section

Meaning that there would already be either a `i2c_stm32_runtime_configure()` or `i2c_stm32_transfer()` on going when this call to `i2c_stm32_runtime_configure()` happens. This is possible but unlikely and I think that the kind of case that should be avoided from applicative side anyway.

OTOH, it is likely that `i2c_stm32_runtime_configure()` happens while counter is 0, triggering the whole machinery of pinctrl configuration/deconfiguration, which it self generates 2 GPIO clock_on/clock_off best case (worst case, `n` GPIO controllers are involved , so `n` time 2 GPIO clock_on/clock_off).

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:78]: > whether a SoC/core goes in to LP-mode would be dictated the timings provided in the DT power-state properties, depending on the actions done here, this may still occur.

Right, but for it to go in LP-mode, it has to be entering idle, which can't be the case until this function call is completed.

File: drivers/i2c/i2c_ll_stm32.c
Comment (knthm) [reply drivers/i2c/i2c_ll_stm32.c:76]: > This is possible but unlikely and I think that the kind of case that should be avoided from applicative side anyway.

It still disrupts/bypasses the runtime get/put pattern in a way that may lead to unintended side effects. This doesn't rub me the right way from an architectural point of view.

> while counter is 0, triggering the whole machinery of pinctrl configuration/deconfiguration, which it self generates 2 GPIO clock_on/clock_off

I agree with you, I'm just wondering if there's a more streamlined approach to this with runtime PM. I doubt this is the only driver where runtime PM actions cascade into other peripheral actions unnecessarily.

File: drivers/i2c/i2c_ll_stm32.c
Comment (knthm) [reply drivers/i2c/i2c_ll_stm32.c:78]: Idle/LP-mode may be triggered from any deeper nested code waiting for a longer period than the residency + exit latency of a LP-mode. I saw this with the calibration steps in the ADC configuration logic.

As the STM32 maintainer, I'd defer to you if you guarantee this won't happen; just wanted to provide a heads-up :)

Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:78]: > Idle/LP-mode may be triggered from any deeper nested code waiting for a longer period than the residency + exit latency of a LP-mode.

Agreed, but I don't see how it can happen if there is still code to execute. Even if you get interrupted by a ahigh prio task that complete and yield, current function execution will be resumed before the kernel can trigger idle.
If this isn't the case, well, I think we'll have much wide and critical issues to deal with.

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:76]: > It still disrupts/bypasses the runtime get/put pattern in a way that may lead to unintended side effects. This doesn't rub me the right way from an architectural point of view.

One case might happen that we're in the middle of `i2c_stm32_runtime_configure()` and `clock_control_on()` was already called while a higher prio thread is calling `i2c_stm32_transfer()`. 
Fortunately we're protected of this case by the `bus_mutex` semaphore, which will ensure `i2c_stm32_runtime_configure()` is completed and `clock_control_off()` called before `i2c_stm32_transfer()` and `pm_device_runtime_get()` happens. So I don't see things goign really bad in any case.

File: drivers/i2c/i2c_ll_stm32.c
Comment (knthm) [reply drivers/i2c/i2c_ll_stm32.c:78]: True, idle will only happen if some code inside waits for any reason. If that's not the case, then there's no purpose for the busy_set logic and you can safely ignore my warning above :)

My though process was that if it was there before it must have some concrete reason, e.g. the I2C configure code waiting for something long enough, but looking at v1/v2 `stm32_i2c_configure_timing()` that doesn't seem to be the case.

File: drivers/i2c/i2c_ll_stm32.c
Comment (knthm) [reply drivers/i2c/i2c_ll_stm32.c:76]: You're right, the `bus_mutex` will always guard this and transactions sections, regardless. From a design point of view it feels like we're breaking the device runtime contract directly toggling peripheral clocks, but there's no concrete reason current why we shouldn't do this.

**Review by knthm — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**
